libraries/persistent_store/src/format.rs

```diff
@@ -0,0 +1,937 @@
+    }
+
+    /// Returns whether a storage is supported.
+    ///
+    /// A storage is supported if the following conditions hold:
+    /// - The size of a word is 4 bytes.
+    /// - The size of a word evenly divides the size of a page.
+    /// - A page contains at least 8 words.
+    /// - A page contains at most [`MAX_PAGE_SIZE`] bytes.
+    /// - There are at least 3 pages.
+    /// - There are at most [`MAX_PAGE_INDEX`]` + 1` pages.
+    /// - A word can be written at least twice between erase cycles.
+    /// - The maximum number of erase cycles is at most [`MAX_ERASE_CYCLE`].
+    ///
+    /// [`MAX_PAGE_SIZE`]: constant.MAX_PAGE_SIZE.html
+    /// [`MAX_PAGE_INDEX`]: constant.MAX_PAGE_INDEX.html
+    /// [`MAX_ERASE_CYCLE`]: constant.MAX_ERASE_CYCLE.html
+    fn is_storage_supported<S: Storage>(storage: &S) -> bool {
+        let word_size = storage.word_size();
+        let page_size = storage.page_size();
+        let num_pages = storage.num_pages();
+        let max_word_writes = storage.max_word_writes();
+        let max_page_erases = storage.max_page_erases();
+        word_size == 4
+            && page_size % word_size == 0
+            && (8 * word_size <= page_size && page_size <= MAX_PAGE_SIZE)
+            && (3 <= num_pages && num_pages <= MAX_PAGE_INDEX + 1)
+            && max_word_writes >= 2
+            && max_page_erases <= MAX_ERASE_CYCLE
+    }
+
+    /// The size of a word in bytes.
+    pub fn word_size(&self) -> usize {
+        WORD_SIZE
+    }
+
+    /// The size of a page in bytes.
+    ///
+    /// We have `32 <= self.page_size() <= MAX_PAGE_SIZE` assuming a word is 4 bytes.
+    pub fn page_size(&self) -> usize {
+        self.page_size
+    }
+
+    /// The number of pages in the storage, denoted by `N`.
+    ///
+    /// We have `3 <= N <= MAX_PAGE_INDEX + 1`.
+    pub fn num_pages(&self) -> usize {
+        self.num_pages
+    }
+
+    /// The maximum page index.
+    ///
+    /// We have `2 <= self.max_page() <= MAX_PAGE_INDEX`.
+    pub fn max_page(&self) -> usize {
+        self.num_pages - 1
+    }
+
+    /// The maximum number of times a page can be erased, denoted by `E`.
+    ///
+    /// We have `E <= MAX_ERASE_CYCLE`.
+    pub fn max_page_erases(&self) -> usize {
+        self.max_page_erases
+    }
+
+    /// The maximum key.
+    pub fn max_key(&self) -> usize {
+        MAX_KEY_INDEX
+    }
+
+    /// The maximum number of updates per transaction.
+    pub fn max_updates(&self) -> usize {
+        MAX_UPDATES
+    }
+
+    /// The size of a virtual page in words, denoted by `Q`.
+    ///
+    /// A virtual page is stored in a physical page after the page header.
+    ///
+    /// We have `6 <= Q <= MAX_VIRT_PAGE_SIZE`.
+    pub fn virt_page_size(&self) -> usize {
+        self.page_size() / self.word_size() - CONTENT_WORD
+    }
+
+    /// The maximum length in bytes of a user payload.
+    ///
+    /// We have `20 <= self.max_value_len() <= MAX_VALUE_LEN` assuming words are 4 bytes.
```

Comment:
Here, wouldn't "We have `5 * self.word_size() <= self.max_value_len() <= MAX_VALUE_LEN`" work?